### PR TITLE
[Gekidou MM-39729] Websocket Events - Groups

### PR DESCRIPTION
**Pull request by jfrerich**
#### Summary
The original primise of this PR was to handle groups websocket events.  After a meeting with Elias and Martink, we decided to scrap the original PR and comment out all of the Groups related code. (see https://community-daily.mattermost.com/core/pl/o859xokph7b6zxcpwnzhnthpuh)

This is a summary of the files that were affected in this PR.

* Table creation
  * any functions that reference `GroupModel`, `GroupMembershipModel`, `GroupsChannelModel`, `GroupsTeamModel`
* Component consumers of Groups models
  *  `app/components/markdown/at_mention/index.tsx`
  * `app/components/post_draft/send_handler`
  * `app/components/post_list/post/body/message`
* server fetch functions
  * `app/actions/remote/group.ts`
* comparators
  * `app/database/operator/server_data_operator/comparators/index.ts`
* handlers 
    * `app/database/operator/server_data_operator/handlers/group.ts`
* Models
  * group models `app/database/models/server/`
    * `group.ts`
    * `group_membership.ts`
    * `groups_channel.ts`
    * `groups_team.ts`
  * group association fields
    * `team.ts` - groupsTeam id
    * `user.ts` - groupMembership
    * `channel.ts` - groupsChannel
* transformers
  * `app/database/operator/server_data_operator/transformers/group.ts`
* table_schemes
  * `app/database/schema/server/table_schemes`
    * `group.ts`
    * `group_membership.ts`
    * `groups_channel.ts`
    * `groups_team.ts`
* types
  * `types/database/models/servers/`
    * `group.d.ts`
    * `group_membership.d.ts`
    * `groups_channel.d.ts`
    * `groups_team.d.ts`
    * `user.d.ts`
    * `channel.d.ts`
    * `team.d.ts`
  * `types/database/raw_values.d.ts`
    * remove `GroupsChannelRelation`
    * remove `GroupsTeamRelation`
* Utils functions
  * `app/utils/draft/index.ts`
  * `app/utils/post/index.ts`
  * `app/utils/user/index.ts`

##### Key points for this PR:
* only add records to tables if needed for the current user.
   * tables  `GroupsChannel`, `GroupsTeam`, `Group`, `GroupMemberships`
* how to achieve this.
  * Group Membership WS received
    * This is the only WS event that will fetch if needed for `Group`, `GroupsTeam`, `GroupsChannel` tables.
    * compare CurrentUserId to incoming WS GroupMembership user_id
    * if not a match early exit
    * if matches, fetch `Group`, `GroupsTeam`, `GroupsChannel` if needed. (similar to roles)
* for other WS events, check if record exists with provided group_id. 
  * if not return
  * if yes, update it
* if user is removed from GroupMembership..
  * remove all group_id entries from all tables??

##### WS events
* `RECEIVED_GROUP: 'received_group'`
  * `GROUP` Table updated
* `RECEIVED_GROUP_ASSOCIATED_TO_TEAM: 'received_group_associated_to_team'`
  * `GROUPS_TEAM` table create/update
* `RECEIVED_GROUP_NOT_ASSOCIATED_TO_TEAM: 'received_group_not_associated_to_team'`
  * `GROUPS_TEAM` table remove group_id/team_id
* `RECEIVED_GROUP_ASSOCIATED_TO_CHANNEL: 'received_group_associated_to_channel'`
  * `GROUPS_CHANNEL` table create/update
* `RECEIVED_GROUP_NOT_ASSOCIATED_TO_CHANNEL: 'received_group_not_associated_to_channel'`
  * `GROUPS_CHANNEL` table remove group_id/channel_Id

- [ ] GROUP_MEMBERSHIP needed

### Add a user to a group
* select cn record from `testgroups`
* select group in LHS viewer 
* in RHS, right-click, add attribute
* add `uniqueMember` for one desired `testusers`

**Example this `tgroup` has two members** 

### To login with LDAP user
* Example **firstloginuser.two**
  *  Apache Directory Studio
  * set `userPassword` for a `testuser` ( **jasonldap** for testing)
  * sync ldap in system console
  * login with 
    * username: `uid` (**firstloginuser.two**)
    * password: `userPassword` (**jasonldap**)

### After login as a user

UserId and GroupId added to GroupMembers Table 

Group is added to the UserGroup Table

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-41584

#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Groups
Mismatch in .png table names for groups
image
image

#### Release Note

```release-note

```

### CODE REVIEW

File: app/screens/post_options/components/options/copy_permalink_option/index.tsx [hunk 1]
```diff
@@ -4,7 +4,7 @@
 import {withDatabase} from '@nozbe/watermelondb/DatabaseProvider';
 import withObservables from '@nozbe/with-observables';
 import {combineLatest, of as of$} from 'rxjs';
-import {switchMap} from 'rxjs/dist/types/operators';
```
Comment (jfrerich) [inline app/screens/post_options/components/options/copy_permalink_option/index.tsx:7]: Not related to this PR, but needed do get npm start to compile

File: app/components/post_list/post/body/message/message.tsx [hunk 2]
```diff
@@ -61,9 +63,13 @@ const Message = ({currentUser, groupsForPosts, highlight, isEdited, isPendingOrF
...
     const blockStyles = getMarkdownBlockStyles(theme);
     const textStyles = getMarkdownTextStyles(theme);
+
+    // groups: MM-41882 const mentionKeys = useMemo(() => {
+    // groups: MM-41882     return getMentionKeysForPost(currentUser, post, groupsForPosts);
+    // groups: MM-41882 }, [currentUser, post.message, groupsForPosts]);
     const mentionKeys = useMemo(() => {
-        return getMentionKeysForPost(currentUser, post, groupsForPosts);
-    }, [currentUser, post.message, groupsForPosts]);
+        return getMentionKeysForPost(currentUser);
```
Comment (enahum) [inline app/components/post_list/post/body/message/message.tsx:71]:
```suggestion
        return currentUser.mentionKeys;
```

Comment (enahum) [reply app/components/post_list/post/body/message/message.tsx:71]: I'm thinking that we should also remove the function `getMentionKeysForPost`

**Review by enahum — CHANGES_REQUESTED** [review]
a couple of changes. (FYI have not reviewed everything)

This is the important part, I see you commented everything related to groups and I know that is what we discussed, but now that I'm looking at it I wonder if we should just remove the code related to groups. Why? The current groups related code might change and I don't know how much, so perhaps best to leave it to @mkraft team to add the feature in full and not get distracted by all this commented code.

What do you guys think?

**Comment by jfrerich** [comment]
> a couple of changes. (FYI have not reviewed everything)
> 
> This is the important part, I see you commented everything related to groups and I know that is what we discussed, but now that I'm looking at it I wonder if we should just remove the code related to groups. Why? The current groups related code might change and I don't know how much, so perhaps best to leave it to @mkraft team to add the feature in full and not get distracted by all this commented code.
> 
> What do you guys think?

It is not going to hurt my feelings to not get a PR merged with a bunch of comments.  I believe the most important responsibility of the is PR should be removing/commenting the code that creates the groups tables. (and this is if we are going to push this out before getting groups fully implemented)

I could also close this PR and reference it in the Jira issue responsible for implementing Groups.

**Comment by enahum** [comment]
> > a couple of changes. (FYI have not reviewed everything)
> > This is the important part, I see you commented everything related to groups and I know that is what we discussed, but now that I'm looking at it I wonder if we should just remove the code related to groups. Why? The current groups related code might change and I don't know how much, so perhaps best to leave it to @mkraft team to add the feature in full and not get distracted by all this commented code.
> > What do you guys think?
> 
> It is not going to hurt my feelings to not get a PR merged with a bunch of comments. I believe the most important responsibility of the is PR should be removing/commenting the code that creates the groups tables. (and this is if we are going to push this out before getting groups fully implemented)
> 
> I could also close this PR and reference it in the Jira issue responsible for implementing Groups.

Ok @jfrerich then instead of commenting everything, let's just delete the code related to groups

File: app/components/markdown/at_mention/index.tsx [hunk 3]
```diff
@@ -99,13 +104,17 @@ const AtMention = ({
...
         }
-        return getUserMentionKeys(user, groups, myGroups);
-    }, [currentUserId, groups, mentionKeys, myGroups, user]);
 
-    const getGroupFromMentionName = () => {
-        const mentionNameTrimmed = mentionName.toLowerCase().replace(/[._-]*$/, '');
-        return groups.find((g) => g.name === mentionNameTrimmed);
-    };
+        // groups: MM-41882 return getUserMentionKeys(user, groups, myGroups);
+        return getUserMentionKeys(user);
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply app/components/markdown/at_mention/index.tsx]: done.

File: app/components/post_list/post/body/message/message.tsx (hunk 2 above)
Comment (jfrerich) [reply app/components/post_list/post/body/message/message.tsx]: also done.

**Comment by jfrerich** [comment]
@enahum I removed all that is client side regarding groups except for the server API types `types/api/groups.d.ts` and rest class for ClientGroups. (app/client/rest/groups.ts).

File: app/components/post_list/post/body/message/index.ts [hunk 4]
```diff
@@ -22,15 +20,12 @@ type MessageInputArgs = {
     post: PostModel;
 }
 
-const withMessageInput = withObservables(['post'], ({database, post}: WithDatabaseArgs & MessageInputArgs) => {
+const withMessageInput = withObservables(['post'], ({database}: WithDatabaseArgs & MessageInputArgs) => {
```
Comment (enahum) [inline app/components/post_list/post/body/message/index.ts:23]:
```suggestion
const withMessageInput = withObservables([], ({database}: WithDatabaseArgs) => {
```

**Review by enahum — CHANGES_REQUESTED** [review]
one small change and do run `npm run i18n-extract` to remove everything group related from the `en.json` translation file

**Pull request merged**